src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -717,14 +717,15 @@ fn compute_deps_doc(
                 unit_for.root_compile_kind(),
             );
             deps_of(scrape_unit, state, unit_for)?;
-            ret.push(new_unit_dep(
+            ret.push(new_unit_dep_with_profile(
```

Comment:
In trying to better understand when someone chooses to use `new_unit_dep` instead of `new_unit_dep_with_profile`, I was lost until I found a comment for the only direct use of `new_unit_dep_with_profile`.

Maybe this deserves a similar comment
(granted, I know you aren't the one who made the change so unsure if you feel confident in doing so)

Reply:
Everything should use `new_unit_dep` except for the one place it is done for build scripts because build scripts need different profiles for build-overrides.

This change doesn't appear correct to me.  The line just above (`let unit_for = …`) that shadows `unit_for` does not seem like the right approach.  It is generating an incorrect `unit_for` with the wrong settings, and that is why the rest of the units end up with the wrong profile.  Commenting out that line seems to get all tests to pass.

Perhaps @willcrichton can say why it is creating a new `unit_for` instead of inheriting the one from the doc unit.

Reply:
Ok after some inspection / experimentation, I believe that @ehuss is right and my code was incorrect. The original intention was to fix an issue that I just realized had a deeper root cause I was able to identify and fix (that proc macros should never actually be scraped). I will put up my own PR so @lqd doesn't have to do more work on my behalf.